cpp/include/cudf/io/types.hpp

```diff
@@ -80,6 +82,14 @@ enum statistics_freq {
   STATISTICS_PAGE = 2,     //!< Per-page column statistics
 };
 
+/**
+ * @brief Table metadata for io readers/writers (primarily column names)
+ */
+struct table_metadata {
+  std::vector<std::string> column_names; //!< Names of columns contained in the table
+  std::map<std::string, std::string> user_data; //!< Format-dependent metadata as key-values pairs
```

Comment:
What sort of metadata is needed?

Reply:
The table-level metadata (usually ignored, but may become useful in the future) typically contains writer-specific information: pandas and Spark store a JSON string containing detailed schema information (like numpy types), arrow stores proprietary data that looks like binary encoded as text (not sure what it is). In all cases I've seen , the map actually only contains one big string.

Reply:
For avro/orc/parquet), that is: for CSV it will always be empty (unless we want to use it to return things like comments).

Reply:
For example, with Spark:
`key="org.apache.spark.sql.parquet.row.metadata", value="{"type":"struct","fields":[{"name":"c","type":"timestamp","nullable":true,"metadata":{}}]}"`
(largely redundant with the standard parquet information, but may allow an app to store extra info, or just keep it unchanged if saving back the file with the same schema)